### DISCUSSION TITLE
Initial Setup Trouble

### DISCUSSION THREAD

**Discussion by ghost** [AdhocAdam/smletsexchangeconnector#317] [Q&A] [answered]
I'm trying to setup the newest release of the connector in my dev environment. I was using an older version of the connector with impersonation. When I run the connector manually via the ISE I get this error:

```
Method invocation failed because [Microsoft.Exchange.WebServices.Data.ImpersonatedUserId] does not contain a method named 'new'.
At E:\PowerShell\Libraries\smletsexchangeconnector-3.3.1.21\smletsExchangeConnector.ps1:4163 char:9
+         $exchangeService.ImpersonatedUserId = [Microsoft.Exchange.WebServices.Da ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : MethodNotFound
```

Also, this is the error in Event Viewer:

```
System.Management.Automation.MethodInvocationException: Exception calling "Bind" with "2" argument(s): "The request failed. The remote server returned an error: (401) Unauthorized." ---> Microsoft.Exchange.WebServices.Data.ServiceRequestException: The request failed. The remote server returned an error: (401) Unauthorized. ---> System.Net.WebException: The remote server returned an error: (401) Unauthorized.
   at System.Net.HttpWebRequest.GetResponse()
   at Microsoft.Exchange.WebServices.Data.ServiceRequestBase.GetEwsHttpWebResponse(HttpWebRequest request)
   --- End of inner exception stack trace ---
   at Microsoft.Exchange.WebServices.Data.ServiceRequestBase.GetEwsHttpWebResponse(HttpWebRequest request)
   at Microsoft.Exchange.WebServices.Data.MultiResponseServiceRequest`1.Execute()
   at Microsoft.Exchange.WebServices.Data.ExchangeService.BindToFolder(FolderId folderId, PropertySet propertySet)
   at Microsoft.Exchange.WebServices.Data.ExchangeService.BindToFolder[TFolder](FolderId folderId, PropertySet propertySet)
   at CallSite.Target(Closure , CallSite , Type , Object , Object )
   --- End of inner exception stack trace ---
   at System.Management.Automation.ExceptionHandlingOps.CheckActionPreference(FunctionContext funcContext, Exception exception)
   at System.Management.Automation.Interpreter.ActionCallInstruction`2.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
```

If I turn off impersonation the script runs in the ISE without error but I still get the error above in event viewer.

**Comment by AdhocAdam** [comment]
I was never able to reproduce this, but this IDENTICAL error has come up before on [this discussion.](https://github.com/AdhocAdam/smletsexchangeconnector/discussions/259)

Per the author:
> **Update** I have continued working though out the day and I was able to narrow down the first issue (Impersonation) to either a bad password on a run-as account or the use of impersonation. I created a new run-as account for the new mailbox and have been able to successful get new connector to work.

Unfortunately I don't have much in the way of troubleshooting as EWS v1.2 and v2.2 both offer the "new" method per the ISE screenshot in that thread. So the only assumption I can make here is it's either the password of the account OR the Run As Account you are using doesn't have Exchange Impersonation rights. But I am certainly open to working through this together and discovering a third possibility!

![ewsImpersonate](https://user-images.githubusercontent.com/6636040/131384562-6590829a-385a-448d-880e-35aed3bbcb40.PNG)

**Comment by ghost** [comment]
![image](https://user-images.githubusercontent.com/36008255/131873329-c1870bbe-6ce6-43bf-b3b2-4a22f3c98e67.png)

When I use the code on that other thread to look for the new method, this is what I get.

**Reply by AdhocAdam** [reply]
![what02](https://user-images.githubusercontent.com/6636040/131878057-0d97f968-9fa7-41e7-b3bd-eedf84392935.gif)

**Comment by ghost** [comment, marked as answer]
Should I see an option for Exchange at this point:

![image](https://user-images.githubusercontent.com/36008255/131873518-0d4c545d-725b-4cd6-b716-62cd02eb2b04.png)

**Reply by AdhocAdam** [reply]
I've found I need to type out at least 

```powershell
[microsoft.exchange.webservices.
```

before intellisense picks up Exchange namespaces

![ewsnamespace](https://user-images.githubusercontent.com/6636040/131879961-b19dd360-e428-40ed-8904-6a19d51ee2cf.png)

**Reply by ghost** [reply]
Ok, yes, it's the same for me.

**Reply by AdhocAdam** [reply]
I did a really quick test - the new method appears for me on EWS v1.2, v2.2, PowerShell ISE, _and_ PowerShell ISE x86.

🤔

**Reply by ghost** [reply]
I've tried on several different servers. My exchange guy tried right on the exchange server itself. We aren't able to get to the new method anywhere. Does version 1.4.4 utilize that too? I have that running in my production environment.

**Reply by AdhocAdam** [reply]
No. Because v1 and v2 of the SMLets Exchange Connector didn't utilize Run As Accounts. In these cases you always had to get credentials to the script somehow. Since people would configure this through Task Scheduler, the task was given the credentials, and then used Windows Auth to talk to Exchange. Thus, official Exchange Impersonation never really came into question because you were always providing the correct credentials for a given mailbox.

With the introduction of v3, support for Run As Accounts were introduced. Since SCSM was now starting the script, it could ANY pass Run As Account creds into it. In this case, there was a need for official Impersonation support as the Run As Account could theoretically be different than the mailbox it was connecting to.

BUT! It still doesn't explain how this method doesn't exist for you.

**Reply by ghost** [reply]
If the run as account is the same as the mailbox account, can I skip using impersonation? If I turn impersonation off the powershell script runs without error; but I still get the 401 unauthorized error in event viewer.

**Reply by AdhocAdam** [reply]
Yes - if the Run As Account is the same as the mailbox you do not need to use Impersonation. Most folks use Impersonation when they want to keep the Mailbox separate from a Run As Account. The security logic being, if you were to compromise the SCSM mailbox you _would not_ compromise an SCSM Run As Account.

To troubleshoot authentication, if you navigate to the EWS endpoint from the SCSM workflow server and enter the Run As Creds does it work?

**Reply by ghost** [reply]
Yes. I get a page that says service at the top and gives some instructions

**Reply by AdhocAdam** [reply]
If you configure v3 to how you've configured v1 (of course configuring the Settings UI first) - do you still get an unauthorized error?

**Reply by ghost** [reply]
If I do that (with Impersonation enabled) the powershell script errors out AND I get this: 

`System.Management.Automation.MethodInvocationException: Exception calling "Bind" with "2" argument(s): "The request failed. The remote server returned an error: (401) Unauthorized." ---> Microsoft.Exchange.WebServices.Data.ServiceRequestException: The request failed. The remote server returned an error: (401) Unauthorized. ---> System.Net.WebException: The remote server returned an error: (401) Unauthorized.
   at System.Net.HttpWebRequest.GetResponse()
   at Microsoft.Exchange.WebServices.Data.EwsHttpWebRequest.Microsoft.Exchange.WebServices.Data.IEwsHttpWebRequest.GetResponse() in \\REDMOND\EXCHANGE\BUILD\E15\15.00.0913.015\SOURCES\sources\dev\EwsManagedApi\src\EwsManagedApi\Core\EwsHttpWebRequest.cs:line 113
   at Microsoft.Exchange.WebServices.Data.ServiceRequestBase.GetEwsHttpWebResponse(IEwsHttpWebRequest request) in \\REDMOND\EXCHANGE\BUILD\E15\15.00.0913.015\SOURCES\sources\dev\EwsManagedApi\src\EwsManagedApi\Core\Requests\ServiceRequestBase.cs:line 821
   --- End of inner exception stack trace ---
   at Microsoft.Exchange.WebServices.Data.ServiceRequestBase.GetEwsHttpWebResponse(IEwsHttpWebRequest request) in \\REDMOND\EXCHANGE\BUILD\E15\15.00.0913.015\SOURCES\sources\dev\EwsManagedApi\src\EwsManagedApi\Core\Requests\ServiceRequestBase.cs:line 831
   at Microsoft.Exchange.WebServices.Data.ServiceRequestBase.ValidateAndEmitRequest(IEwsHttpWebRequest& request) in \\REDMOND\EXCHANGE\BUILD\E15\15.00.0913.015\SOURCES\sources\dev\EwsManagedApi\src\EwsManagedApi\Core\Requests\ServiceRequestBase.cs:line 724
   at Microsoft.Exchange.WebServices.Data.MultiResponseServiceRequest`1.Execute() in \\REDMOND\EXCHANGE\BUILD\E15\15.00.0913.015\SOURCES\sources\dev\EwsManagedApi\src\EwsManagedApi\Core\Requests\MultiResponseServiceRequest.cs:line 157
   at Microsoft.Exchange.WebServices.Data.ExchangeService.BindToFolder(FolderId folderId, PropertySet propertySet) in \\REDMOND\EXCHANGE\BUILD\E15\15.00.0913.015\SOURCES\sources\dev\EwsManagedApi\src\EwsManagedApi\Core\ExchangeService.cs:line 278
   at Microsoft.Exchange.WebServices.Data.ExchangeService.BindToFolder[TFolder](FolderId folderId, PropertySet propertySet) in \\REDMOND\EXCHANGE\BUILD\E15\15.00.0913.015\SOURCES\sources\dev\EwsManagedApi\src\EwsManagedApi\Core\ExchangeService.cs:line 295
   at CallSite.Target(Closure , CallSite , RuntimeType , Object , Object )
   --- End of inner exception stack trace ---
   at System.Management.Automation.ExceptionHandlingOps.CheckActionPreference(FunctionContext funcContext, Exception exception)
   at System.Management.Automation.Interpreter.ActionCallInstruction`2.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)`

**Reply by ghost** [reply]
If I disable impersonation I don't get a powershell error, but I still get the event log error I just posted. Also, the mailbox is not processed.

**Reply by AdhocAdam** [reply]
I see Redmond/Exchange - is this Exchange Online? or On Premise?

**Reply by ghost** [reply]
It's On Premise.

**Reply by ghost** [reply]
I guess there is something funky with the exchange server certificate and the autodiscover url; so we're wondering if that may be causing an issue even though we aren't using the autodiscover in the setup. I have the endpoint url instead: https://webmail.oursite.com/EWS/Exchange.asmx

**Reply by AdhocAdam** [reply]
Yea the next test I was going to suggest is trying to retrieve even just your own inbox with a manual configuration. This would either prove the issue is with the account OR Exchange itself.

https://adhocadam.github.io/smletsexchangeconnector/the-smlets-exchange-connector-at-its-core/

**Reply by ghost** [reply]
Using that script I can use autodiscover or not use it and it works greats. I get logged in and retrieve both emails from the inbox.

**Reply by AdhocAdam** [reply]
Is the email address the same email address that is being used for the selected account?

![screen](https://user-images.githubusercontent.com/6636040/131895681-e70dd283-386a-46c6-ad3a-9680d406faa2.PNG)

**Reply by ghost** [reply]
Yes.

**Reply by AdhocAdam** [reply]
So the testing script works but the connector + run as accounts dont. Hrmmmm...

Is the Run As Account's username formatted as **domain.tld\username** ? and/or how were you starting PowerShell with the tester script that it worked?

**Reply by ghost** [reply]
In the run as account the username is just 'username' the domain is defined lower down:
![image](https://user-images.githubusercontent.com/36008255/131901686-c48d774d-9e47-47dd-a74e-20476d324e87.png)

In the test script it is the same, just 'username' with the domain defined separately:
![image](https://user-images.githubusercontent.com/36008255/131901571-4265eee2-2004-4d74-9f93-8b76bc139796.png)

The 1.4.4 script works fine using impersonation and the account defined in the script:
![image](https://user-images.githubusercontent.com/36008255/131901925-8a3abe6c-60c3-45e5-b6cc-44b70f1d8f9b.png)

**Reply by AdhocAdam** [reply]
I'm still trying to figure out a way to reproduce this in v3 but haven't been successful so far. I should add that if I choose Impersonation, the error I get cites an inability to Impersonate rather than Authenticate.

**Impersonation Error:**
 "The account does not have permission to impersonate the requested user." ---> Microsoft.Exchange.WebServices.Data.ServiceResponseException: The account does not have permission to impersonate the requested user.

**Reply by AdhocAdam** [reply]
The only way I've managed to reproduce this is:
- The workflow is turned off
- I'm running the script manually (the account does not matter)
- I've manually edited $exchangeAuthenticationType to equal "impersonation"

Essentially, the ELSE -> SWITCH in the primary Exchange Authentication logic is engaged and is expecting credentials but since they aren't defined anywhere they all evaluate to null.
https://github.com/AdhocAdam/smletsexchangeconnector/blob/4740de699b759a09f2af15081751f0ba3e67ddc0/smletsExchangeConnector.ps1#L4166-L4205

The attempt to authenticate is made with null credentials, an error is thrown and then written to the log.
https://github.com/AdhocAdam/smletsexchangeconnector/blob/4740de699b759a09f2af15081751f0ba3e67ddc0/smletsExchangeConnector.ps1#L4219-L4227

But the moment the Workflow is turned on (and thus SCSM is running this automatically) it's resolved.

**Reply by ghost** [reply]
@AdhocAdam - Thank you for helping me with the custom script path and enabling the workflow. That cleared this up for me.
